File: src/ados/sdk/drivers/gps.py
"""GPS driver base class.

A GPS driver decodes a position fix stream from a u-blox, NMEA, RTK,
or vendor-custom receiver and exposes it as a series of :class:`GpsFix`
samples. Drivers that support RTK can also accept RTCM correction
payloads via :meth:`GpsDriver.inject_rtcm`.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass
from typing import Any, AsyncIterator


@dataclass(frozen=True)
class GpsCandidate:
    """A GPS receiver a driver claims it can open."""

    driver_id: str
    device_id: str
    label: str
    bus: str
    vid_pid: tuple[int, int] | None = None
    metadata: dict[str, Any] | None = None


@dataclass(frozen=True)
class GpsCapabilities:
    """Static capabilities of an open GPS session."""

    protocol: str
    constellations: list[str]
    max_update_hz: float
    supports_rtk: bool = False
    supports_dual_band: bool = False
    supports_heading: bool = False


@dataclass(frozen=True)
class GpsFix:
    """One position fix yielded by a GPS driver.

    Latitude and longitude are in degrees (WGS-84). Altitude is metres
    above mean sea level. ``fix_type`` follows the common convention:
    0 no fix, 2 2D, 3 3D, 4 DGPS, 5 RTK float, 6 RTK fixed.
    """

    timestamp_ns: int
    latitude_deg: float
    longitude_deg: float
    altitude_msl_m: float
    fix_type: int
    satellites_used: int
    hdop: float
    vdop: float
    horizontal_accuracy_m: float | None = None
    vertical_accuracy_m: float | None = None
    speed_mps: float | None = None
    course_deg: float | None = None
    heading_deg: float | None = None
    metadata: dict[str, Any] | None = None


class GpsSession:
    """Opaque per-open state held by a GPS driver."""


class GpsDriver(ABC):
    """Abstract base for GPS drivers."""

    @abstractmethod
    async def discover(self) -> list[GpsCandidate]:
        """Scan for GPS receivers this driver can open."""

    @abstractmethod
    async def open(
        self, candidate: GpsCandidate, config: dict[str, Any]
    ) -> GpsSession:
        """Open a session against a GPS receiver."""

    @abstractmethod
    async def close(self, session: GpsSession) -> None:
        """Release resources held by a session."""

    @abstractmethod
    def capabilities(self, session: GpsSession) -> GpsCapabilities:
        """Return the static capabilities of an open session."""

    @abstractmethod
    async def fix_iterator(self, session: GpsSession) -> AsyncIterator[GpsFix]:
        """Yield fixes until the session closes."""

    @abstractmethod
    async def inject_rtcm(self, session: GpsSession, payload: bytes) -> None:
        """Forward an RTCM correction packet into the receiver.

        Receivers without RTK support should raise :class:`NotImplementedError`.
        """

File: crates/ados-sdk/src/drivers/gps.rs
//! GPS driver trait.
//!
//! Ports `ados.sdk.drivers.gps`. A GPS driver decodes a position-fix stream
//! from a u-blox, NMEA, RTK, or vendor-custom receiver and exposes it as a
//! series of [`GpsFix`] samples. RTK-capable drivers accept RTCM corrections
//! via [`GpsDriver::inject_rtcm`].

use std::collections::BTreeMap;

use async_trait::async_trait;
use rmpv::Value;

use super::{Bus, DriverResult, SampleStream};

/// A GPS receiver a driver claims it can open.
#[derive(Debug, Clone, PartialEq)]
pub struct GpsCandidate {
    pub driver_id: String,
    pub device_id: String,
    pub label: String,
    pub bus: Bus,
    pub vid_pid: Option<(u16, u16)>,
    pub metadata: BTreeMap<String, Value>,
}

/// Static capabilities of an open GPS session.
#[derive(Debug, Clone, PartialEq)]
pub struct GpsCapabilities {
    pub protocol: String,
    pub constellations: Vec<String>,
    pub max_update_hz: f32,
    pub supports_rtk: bool,
    pub supports_dual_band: bool,
    pub supports_heading: bool,
}

/// One position fix. Latitude/longitude are degrees (WGS-84), altitude metres
/// above mean sea level. `fix_type`: 0 none, 2 2D, 3 3D, 4 DGPS, 5 RTK float,
/// 6 RTK fixed.
#[derive(Debug, Clone, PartialEq)]
pub struct GpsFix {
    pub timestamp_ns: i64,
    pub latitude_deg: f64,
    pub longitude_deg: f64,
    pub altitude_msl_m: f32,
    pub fix_type: u8,
    pub satellites_used: u8,
    pub hdop: f32,
    pub vdop: f32,
    pub horizontal_accuracy_m: Option<f32>,
    pub vertical_accuracy_m: Option<f32>,
    pub speed_mps: Option<f32>,
    pub course_deg: Option<f32>,
    pub heading_deg: Option<f32>,
    pub metadata: BTreeMap<String, Value>,
}

/// Abstract GPS driver.
#[async_trait]
pub trait GpsDriver: Send + Sync {
    type Session: Send + Sync;

    /// Scan for GPS receivers this driver can open.
    async fn discover(&self) -> DriverResult<Vec<GpsCandidate>>;

    /// Open a session against a GPS receiver.
    async fn open(
        &self,
        candidate: &GpsCandidate,
        config: &BTreeMap<String, Value>,
    ) -> DriverResult<Self::Session>;

    /// Release resources held by a session.
    async fn close(&self, session: Self::Session) -> DriverResult<()>;

    /// Return the static capabilities of an open session.
    fn capabilities(&self, session: &Self::Session) -> GpsCapabilities;

    /// Yield fixes until the session closes.
    async fn fix_iterator(&self, session: &Self::Session) -> DriverResult<SampleStream<GpsFix>>;

    /// Forward an RTCM correction packet into the receiver. Receivers without
    /// RTK return [`super::DriverError::NotSupported`].
    async fn inject_rtcm(&self, session: &Self::Session, payload: &[u8]) -> DriverResult<()>;
}
